### PR TITLE
Provide fast generic comparer for bool values

### PR DESCRIPTION
**Pull request by vasily-kirichenko**

### CODE REVIEW

**Comment by dsyme** [comment]
@dotnet-bot Test Ubuntu16.04 Release_default Build please

**Comment by dsyme** [comment]
@dotnet-bot Test Ubuntu16.04 Release_fcs Build please

**Comment by smoothdeveloper** [comment]
@dsyme I've noticed the ubuntu builds tend to fail due to nuget pulling packages.

Since nuget tend to fail as a nuget client when server is busy, I think it would make sense to use paket for the same role, it retries the download few times and generally complete with higher success rate AFAIU.

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 1]
```diff
@@ -2218,6 +2220,7 @@ namespace Microsoft.FSharp.Core
                 | ty when ty.Equals(typeof<float32>)    -> unboxPrim (box Float32Comparer)
                 | ty when ty.Equals(typeof<decimal>)    -> unboxPrim (box DecimalComparer)
                 | ty when ty.Equals(typeof<string>)     -> unboxPrim (box StringComparer)
+                | ty when ty.Equals(typeof<bool>)     -> unboxPrim (box BoolComparer)
```
Comment (smoothdeveloper) [inline src/fsharp/FSharp.Core/prim-types.fs]: nit: that code was vertically aligned, we should keep it aligned.

I'm baffled there are specialized cases for many types but not for bool!

**Comment by cartermp** [comment]
@KevinRansom @TIHan @brettfo Assuming this is approved by Monday, let's aim to merge this into 15.8 and F# 4.5.

**Comment by dsyme** [comment]
> Looks good to me. Although, I have to admit, that sorting an array of bools, might not be the most common algorithm in computer science.

The big uses likely happen when bools are used as part of some larger key, e.g. a dictionary keyed by `(string * bool)` or `type Key = Key of string * bool`.  Even in these cases the tuple comparer should normally get inlined and specialised but in cases where this doesn't happen for some reason the key comparisons may end up going down the routes in this PR.  (as an aside I'd be curious of the source of the above comparisons and why specialization wasn't kicking in)

**Pull request merged**